File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        cin.tie(0);
        cout.tie(0);
        ios::sync_with_stdio(false);
        vector<vector<int>> res;
        permuterec(nums,0,res);
        return res;
    }
    vector<vector<int>> permuterec(vector<int>&nums,int ci,vector<vector<int>>&res){
        if(ci==nums.size()-1){
            res.push_back(nums);
            return res;
        }
        for(int i=ci;i<nums.size();i++){
            swap(nums[ci],nums[i]);
            permuterec(nums,ci+1,res);
            swap(nums[ci],nums[i]);
        }
        return res;
    }
};

File: 0046-permutations/0046-permutations.java
class Solution {
    void permute(int[]nums,int l,int r,List<List<Integer>>res){
        if(l==r){
            List<Integer> list = Arrays.stream(nums).boxed().collect(Collectors.toList());
            res.add(list);
            return;
        }
        for(int i=l;i<=r;i++){
            swap(nums,l,i);
            permute(nums,l+1,r,res);
            swap(nums,l,i);
        }
    }
    void swap(int []nums,int l,int r){
        int temp=nums[l];
        nums[l]=nums[r];
        nums[r]=temp;
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>>res=new ArrayList<>();
        permute(nums,0,nums.length-1,res);
        return res;
    }
}